// node-db-migration/lib/drivers/pgsql.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PsqlDriver = void 0;
const abstractdriver_1 = require("./abstractdriver");
class PsqlDriver extends abstractdriver_1.CommonDriver {
    isInitedSql() {
        return `SELECT 1 FROM information_schema.tables WHERE table_name = '${this.migrationTable}'`;
    }
    getSeparator() {
        let i = 0;
        return () => {
            i++;
            return `$${i}`;
        };
    }
    createTableSql() {
        return `CREATE TABLE ${this.migrationTable}` +
            `(` +
            `    id bigserial PRIMARY KEY ,` +
            `    name VARCHAR(128) NOT NULL,` +
            `    run_on TIMESTAMP DEFAULT CURRENT_TIMESTAMP NOT NULL,` +
            `    created TIMESTAMP DEFAULT CURRENT_TIMESTAMP NOT NULL,` +
            `    error_if_happened text` +
            `)`;
    }
    readQuery(sql, params) {
        return __awaiter(this, void 0, void 0, function* () {
            return new Promise((resolve) => {
                this.dbRunner.query(sql, params, (err, result) => {
                    const newVar = {
                        error: this.extractError(err),
                        rows: result ? result.rows : [],
                    };
                    resolve(newVar);
                });
            });
        });
    }
}
exports.PsqlDriver = PsqlDriver;
