// node-db-migration/lib/drivers/sqlight.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SQLite3Driver = void 0;
const abstractdriver_1 = require("./abstractdriver");
class SQLite3Driver extends abstractdriver_1.CommonDriver {
    isInitedSql() {
        return `SELECT name FROM sqlite_master WHERE type='table' AND name='${this.migrationTable}'`;
    }
    query(sql, params) {
        return __awaiter(this, void 0, void 0, function* () {
            return new Promise((resolve) => {
                this.dbRunner.run(sql, params, (error) => {
                    resolve(this.extractError(error));
                });
            });
        });
    }
    executeMultipleStatements(sql) {
        return __awaiter(this, void 0, void 0, function* () {
            return new Promise((resolve) => {
                this.dbRunner.exec(sql, (err) => {
                    resolve(this.extractError(err));
                });
            });
        });
    }
    readQuery(sql, params) {
        return __awaiter(this, void 0, void 0, function* () {
            return new Promise((resolve) => {
                this.dbRunner.all(sql, params, (error, result) => {
                    resolve({ error: this.extractError(error), rows: result });
                });
            });
        });
    }
    createTableSql() {
        return `CREATE TABLE ${this.migrationTable}` +
            `(` +
            `    id INTEGER  NOT NULL PRIMARY KEY AUTOINCREMENT ,` +
            `    name VARCHAR(128) NOT NULL,` +
            `    run_on DATETIME DEFAULT CURRENT_TIMESTAMP NOT NULL,` +
            `    created DATETIME DEFAULT CURRENT_TIMESTAMP NOT NULL,` +
            `    error_if_happened text` +
            `)`;
    }
}
exports.SQLite3Driver = SQLite3Driver;
